src/hotspot/os/bsd/os_bsd.cpp

```diff
@@ -2230,11 +2208,47 @@ int os::open(const char *path, int oflag, int mode) {
   // 4843136: (process) pipe file descriptor from Runtime.exec not being closed
   // 6339493: (process) Runtime.exec does not close all file descriptors on Solaris 9
   //
-#ifdef FD_CLOEXEC
+  // Modern Linux kernels (after 2.6.23 2007) support O_CLOEXEC with open().
+  // O_CLOEXEC is preferable to using FD_CLOEXEC on an open file descriptor
+  // because it saves a system call and removes a small window where the flag
+  // is unset.  On ancient Linux kernels the O_CLOEXEC flag will be ignored
+  // and we fall back to using FD_CLOEXEC (see below).
```

Comment:
This comment was confusing me until I looked at os_linux.cpp and see
that you are getting in sync with the comments in that file (which is
where this code was ported from).

Your changes below are also getting in sync with the code in the
os_linux.cpp file.

Did you want to mention here the macOS version info that includes
the support for O_CLOEXEC?

Reply:
Thank you for the review Dan!

I thought about it, but I decided that keeping the code in sync between linux and macOS was the priority. Eventually we might want to have a single impl shared between linux and mac I thought.

I ran Mach5 hs-tier1...5 for the testing.

Reply:
Sorry but copying a comment from os_linux to os_bsd really doesn't make sense. We've been cleaning up these "mis-ports" ever since the macOS port was done, so please don't add new occurrences.

Reply:
I see now it is just not comments but also the code. Again sorry but if the Linux workarounds are for Linux then they don't belong in the BSD code. That just confuses anyone who looks at this code later. If there were eventually an os_posix shared version of this then it wouldn't include Linux specific workarounds.

Reply:
This is not a **workaround** but an improvement to the robustness of the code, that applies equally to both Linux and macOS, by taking an advantage of what once was a new feature, but now is just a more reliable mechanism for making sure that the forked/child process does not get access to unintended "leaked" file descriptor. The original comment came from Linux, but we can change it, to apply equally to both Linux/macOS.

I will refactor this common POSIX code, though they are not 100% exactly the same - Linux uses `open64()`, while macOS uses `open()`.

Reply:
The comment says:

> On ancient Linux kernels the O_CLOEXEC flag will be ignored and we fall back to using FD_CLOEXEC (see below).

so that is a workaround to handle ancient Linux kernels, and such a workaround has no meaning for BSD unless it too has ancient versions that ignored O_CLOEXEC.

That said, how ancient are we talking about? That comment may be irrelevant today given the minimum supported kernel versions we have.

Reply:
I have tested O_CLOEXEC on macOS and it is available even at 10.9 I don't have hardware that can go earlier than that.

So it looks like on macOS we don't need the fallback, however, macOS is not the only BSD OS that we need to consider (?)

Should we leave the fallback for other BSD platforms?

Reply:
Is there anything to suggest any BSD system has ever had this issue? This seems a Linux specific workaround that has no place on macOS or BSD.

Reply:
If you look at _fcntl.h_ you will find:

```
#if __DARWIN_C_LEVEL >= 200809L
#define O_CLOEXEC       0x01000000      /* implicitly set FD_CLOEXEC */
#endif
```

so that suggests that `O_CLOEXEC` was not always available on macOS, so I assume same holds true for other BSD os'es?

We could tweak the BSD code to statically account for the availability of `O_CLOEXEC` with `#ifdef `, but why not re-use Linux code and do it dynamically as needed?

On Linux this accounts for the real possibility of kernel ignoring `O_CLOEXEC` and on BSD the workaround would be a actually a fallback on those platform that plain did not support `O_CLOEXEC` at all.

We could tweak the comment to reflect these assumptions.

Reply:
Hi Gerard,

sorry, I'm with @dholmes-ora on this one. Copying Linux comments like this is not really helpful.

- Completely different kernels (both BSD and Mac), therefore I would skip the "did the kernel ignore my flag?" check. That was a workaround for one very specific Linux kernel bug.
- The `#ifdef O_CLOEXEC` does not really help you with backward compatibility if it is really your aim to be runnable on older OSes. Since this tests on the build machine, kernel on the runtime machine may not understand (hopefully ignore) the numerical value O_CLOEXEC.

Honestly, I would not bother. I would just use open with O_CLOEXEC and be done with it. If someone really really wants to build JDK 20 on ancient MacOS or BSD, then we can add complexity.

Just my 5 cent. 

Cheers, Thomas